src/PJ_vgridshift.c

```diff
@@ -37,25 +44,60 @@ static LPZ reverse_3d(XYZ xyz, PJ *P) {
 
 
 static PJ_COORD forward_4d(PJ_COORD obs, PJ *P) {
-    PJ_COORD point = {{0,0,0,0}};
-    point.xyz = forward_3d (obs.lpz, P);
+    struct pj_opaque_vgridshift *Q = (struct pj_opaque_vgridshift *) P->opaque;
+    PJ_COORD point = obs;
+    if (Q->t_final != 0.0  && Q->t_epoch != 0.0) {
+        if (obs.lpzt.t < Q->t_epoch && Q->t_final > Q->t_epoch)
+            point.xyz = forward_3d (obs.lpz, P);
+    } else {
+        point.xyz = forward_3d (obs.lpz, P);
+    }
     return point;
 }
 
 static PJ_COORD reverse_4d(PJ_COORD obs, PJ *P) {
-    PJ_COORD point;
-    point.lpz = reverse_3d (obs.xyz, P);
+    struct pj_opaque_vgridshift *Q = (struct pj_opaque_vgridshift *) P->opaque;
+    PJ_COORD point = obs;
+    if (Q->t_final != 0.0  && Q->t_epoch != 0.0) {
+        if (obs.lpzt.t < Q->t_epoch && Q->t_final > Q->t_epoch)
+            point.lpz = reverse_3d (obs.xyz, P);
+    } else {
+        point.lpz = reverse_3d (obs.xyz, P);
+    }
     return point;
 }
 
 
 PJ *TRANSFORMATION(vgridshift,0) {
+    struct pj_opaque_vgridshift *Q = pj_calloc (1, sizeof (struct pj_opaque_vgridshift));
+    if (0==Q)
+        return pj_default_destructor (P, ENOMEM);
+    P->opaque = (void *) Q;
 
    if (!pj_param(P->ctx, P->params, "tgrids").i) {
         proj_log_error(P, "vgridshift: +grids parameter missing.");
         return pj_default_destructor(P, PJD_ERR_NO_ARGS);
     }
 
+   if (pj_param(P->ctx, P->params, "tt_final").i) {
+        Q->t_final = pj_param (P->ctx, P->params, "dt_final").f;
```

Comment:
we should likely have a pj_eval_param_time(const char*) function to do the above, since it is repeated above.
localtime() is known on POSIX to be thread-unsafe. It is better to use localtime_r() instead, when available

Reply:
> we should likely have a pj_eval_param_time(const char*) function to do the below processing, since it is repeated above.

Good point. I intend to add similar functionality to `deformation` and possibly `helmert` as well. I will take care of it then.

> localtime() is known on POSIX to be thread-unsafe. It is better to use localtime_r() instead, when available

When can `localtime_r` be expected to be available?

Reply:
I think it is a bit out of scope for this PR. How big a risk is it to include the code as it is now? localtime() is only called once during creation of the `PJ` object.

Reply:
The risk is probably low. A TODO in the code + issue to remember should be good enough

Reply:
Created #1016

Reply:
You can probably significantly reduce (but not eliminate) the risk by saying:

    struct tm cucumber;
    ...
    cucumber = *localtime(...);

Essentially doing a copy of the the `localtime`internal `struct tm`, rather than storing a pointer to it.